### PR TITLE
SONARJS-703 Provide an easy way to unit test secondary issue locations

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java [hunk 1]
```diff
@@ -213,11 +278,46 @@ private void addPreciseLocation(SyntaxTrivia trivia) {
...
+        return expectedIssue;
+      }
+    }
+
+    String format = "Invalid test file: precise secondary location is provided for ID '%s' but no issue is asserted with such ID (line %s)";
+    String missingAssertionMessage = String.format(format, id, line);
+    throw new IllegalStateException(missingAssertionMessage);
+  }
+
   private static void addParams(TestIssue issue, String params) {
```
Comment (SonarTech) [inline javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java]: MAJOR The Cyclomatic Complexity of this method "addParams" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java [hunk 2]
```diff
@@ -213,11 +278,46 @@ private void addPreciseLocation(SyntaxTrivia trivia) {
...
+        int startIndex = text.indexOf("{{") + 2;
+        message = text.substring(startIndex, endIndex);
+        text = text.substring(0, startIndex - 2).trim();
+      }
+
+      issueByID(text, trivia.line()).secondary(message, trivia.line() - 1, startColumn, endColumn);
+    }
+  }
+
+  private TestIssue issueByID(String id, int line) {
```
Comment (ivandalbosco) [inline javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java]: This method (a get method) should only contains the "for" loop. The last 3 lines (the check) should be moved to method addPreciseSecondaryLocations.

File: javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java [hunk 3]
```diff
@@ -47,6 +48,7 @@
 public class JavaScriptCheckVerifier extends SubscriptionVisitorCheck {
 
   private final List<TestIssue> expectedIssues = new ArrayList<>();
+  private final List<SyntaxTrivia> preciseSecondaryLocationComments = new ArrayList<>();
```
Comment (ivandalbosco) [inline javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java]: @vilchik-elena  This is a nice and seemingly quite robust implementation. Test coverage is excellent.

However class JavaScriptCheckVerifier has become a quite long one. This makes it hard for the reader to distinguish clearly between the class' two purposes: a) getting the expected issues and b) comparing the expected issues with the actual issues. I would suggest to split the class in two by:
- deporting activity a) to a new class, say JavaScriptExpectedIssuesCheck, that extends SubscriptionVisitorCheck
- keeping activity b) as the unique contract of class JavaScriptCheckVerifier, by making class JavaScriptCheckVerifier instanciate JavaScriptExpectedIssuesCheck and no longer extend SubscriptionVisitorCheck.
This opens the door to a more balanced implementation with class JavaScriptCheckVerifier having two mirror methods:
- getActualIssues (already exists, no change required): delegates to the JavaScriptCheck
- getExpectedIssues (to be extracted from the beginning of method verify): delegates to a JavaScriptExpectedIssuesCheck

By the way, do these classes really need to have their name starting with "JavaScript"? The name "JavaScriptCheckVerifier" does not bring much as compared to "CheckVerifier".

File: javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java (hunk 2 above)
Comment (vilchik-elena) [inline javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java]: I do not agree, the exception throwing logic is part of this method logic

File: javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java (hunk 3 above)
Comment (vilchik-elena) [inline javascript-checks-testkit/src/main/java/org/sonar/javascript/checks/verifier/JavaScriptCheckVerifier.java]: Thanks, fixed in 727800f50f7d15fdf299082eca23aa99aca498ba

**Pull request merged**